components/base-node.js

```diff
@@ -65,7 +79,32 @@ module.exports = {
        ar.push(item);
      }
      return ar;
-   } 
+   },
+
+   updatePromise: function(def) {
+     return promiseComponent(
 
+         _.defaults(
+             def,
+             { inPorts: _.mapObject(
+                          def.inPorts,
+                          function(port, portName) {
+
+                            return _.defaults(
+                                     port,
+                                     {
+                                       ondata: function ( payload, socketIndex ) {
+                                                 this[portName] = payload;
+                                                 var self = this;
+                                                 return def.onchange.call( this, _.pick( this, _.keys(def.inPorts)));
```

Comment:
I have added the check for all ports having received a state before calling the updater.  Not sure about the EventEmitter API - let's talk about it.